python/GafferArnoldTest/ArnoldRenderTest.py

```diff
@@ -1534,5 +1534,127 @@ def testInstancerEncapsulatePerf( self ) :
 			with GafferTest.TestRunner.PerformanceScope() :
 				s["render"]["task"].execute()
 
+	@GafferTest.TestRunner.PerformanceTestMethod( repeat = 1 )
+	def testInstancerManyPrototypesPerf( self ) :
+
+		s = Gaffer.ScriptNode()
+
+		s["plane"] = GafferScene.Plane()
+		s["plane"]["divisions"].setValue( imath.V2i( 500 ) )
+
+		s["sphere"] = GafferScene.Sphere()
+		s["sphere"]["expression"] = Gaffer.Expression()
+
+		s["pathFilter"] = GafferScene.PathFilter()
+		s["pathFilter"]["paths"].setValue( IECore.StringVectorData( [ '/plane' ] ) )
+
+		s["instancer"] = GafferScene.Instancer()
+		s["instancer"]["in"].setInput( s["plane"]["out"] )
+		s["instancer"]["filter"].setInput( s["pathFilter"]["out"] )
+		s["instancer"]["prototypes"].setInput( s["sphere"]["out"] )
+
+		s["instancer"]["contextVariables"].addChild( GafferScene.Instancer.ContextVariablePlug( "context" ) )
+		s["instancer"]["contextVariables"][0]["name"].setValue( "P" )
+		s["instancer"]["contextVariables"][0]["quantize"].setValue( 0 )
```

Comment:
It's possible that we should add a "fancy 2 stage hashing process specifically for when there are no children", and then it might require a separate test which doesn't hit the optimal path for that - but currently that fancy special case doesn't exist, and it doesn't matter that much from a performance point of view that the context variable isn't used ( or at least, the extra cost of actually using it is all outside the instancer, and isn't what we want to measure in this test ).

If there are children, then it would be very hard to optimize based on the context not being used - we just create sub-Capsules, which are responsible for doing the scene traversal, and if two of those capsules happen to come out identical after doing that traversal ( ... or all of the sub-Capsules ), we don't have a good way of knowing.

If there are no children, and we're not using sub-Capsules, then it could be pretty reasonable to not just use the Context hash for the map, and instead evaluate the hash of the actual object, and use that as a map key for a second cache ... though you'd also have to include the hash of the attributes. Anyway, the second code path would definitely make things more complex, though it's doable, and would probably be a win for cases where users are setting up a lot of contexts that aren't different ... though the extra Gaffer hash evaluations are always going to hurt a fair bit when users are sloppy with contexts.

Reply:
Your reply suggests that this test is intended to exercise `InstancerCapsule::render()`, but it's not currently doing that since `encapsulateInstanceGroups` isn't set. So I think we need to do the following?

- Set `encapsulateInstanceGroups`
- Remove unused `s["sphere"]["expression"]`
- Add a comment explaining why it doesn't matter that the context isn't being used explicitly

Reply:
This test exists mostly for comparison with the encapsulated case, testInstancerManyPrototypesEncapsulatePerf, which I'm pretty sure is setting encapsulate, and showing that encapsulation is a 7X speedup with no context variables, a 7X speedup with context variables that all resolve to the same prototype ( both encap and unencap are 25% slower ), and only a 2.5X speedup when all prototypes use a different context.

To try and clarify, I have added a couple comments, rearranged testInstancerManyPrototypesPerf and testInstancerManyPrototypesEncapsulatePerf to be next to each other, and removed the unused `s["sphere"]["expression"]` which got pasted into a lot of tests.

Reply:
Ah, cool - thanks for the explanation.